x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/utils/log_cluster_shard_failures_esql.ts

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type { EsqlEsqlShardFailure } from '@elastic/elasticsearch/lib/api/types';
+import type { EsqlTable } from '../esql/esql_request';
+import * as i18n from '../translations';
+import type { SearchAfterAndBulkCreateReturnType } from '../types';
+
+export const logClusterShardFailuresEsql = ({
+  response,
+  result,
+}: {
+  response: EsqlTable;
+  result: SearchAfterAndBulkCreateReturnType;
+}) => {
+  const clusters = response?._clusters?.details ?? {};
+  const shardFailures = Object.keys(clusters).reduce<EsqlEsqlShardFailure[]>((acc, cluster) => {
+    const failures = clusters[cluster]?.failures ?? [];
+    return [...acc, ...failures];
```

Comment:
If there are no cluster failures for a given cluster we are still spreading out the accumulated failures. I imagine if there are many clusters failing, that accumulator could grow large. I'd suggest adding logic to update the accumulator directly or only spread the accumulator if there are new failures to be added.